Fix part of #10700: Refactor Object Factories with Dependencies

darksun27 opened the pull request:
## Overview

1. This PR fixes or fixes part of #10700 
2. This PR does the following: This PR refactors the following object factories:
- CollectionNode
- CollectionPlaythrough
- Collection
- EmailDashboardQuery
- EmailDashboardQueryResults
- PlatformParameter
- PlatformParameterFilter
- PlatformParameterRule
- LearnerExplorationSummary
- ReadOnlyStoryNode
- StoryPlaythrough
- UserProfile
- ExplorationStats
- HighBounceRateTask
- SuccessiveIncorrectAnswersTask
- IneffectiveFeedbackLoopTask
- NeedsGuidingResponesTask
- ExplorationTasks
- TopicSummary
- ClassroomData

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

darksun27 committed:
Refactor Collection Node Object Factory

darksun27 committed:
Refactor Collection Object Factory

darksun27 committed:
Refactor Email Query Deashboard and Email Query Dashboard Results Object Factories

darksun27 committed:
Refactor Platform Parameter Filter, Rule Object Factories

darksun27 committed:
Refactor Exploration Stats Object Factories

darksun27 committed:
Refactor LearnerExplorationSummary and UserProfile Object Factories

darksun27 committed:
Refactor ReadOnlyStoryNode and StoryPlaythrough Object Factories

darksun27 committed:
Refactor TopicSummary and ClassroomData Object Factories

darksun27 committed:
Refactor HighBounceRateTask Object Factory

darksun27 committed:
Refactor SuccessiveIncorrectAnswerTask IneffectiveFeedbackLoopTask NeedsGuidingResponseTask and ExplorationTask Object Factories

darksun27 committed:
Merge remote-tracking branch 'upstream/develop' into refactor-object-factories-red

darksun27 committed:
Fix coverage

darksun27 committed:
Fix coverage

darksun27 committed:
Fix coverage

darksun27 committed:
Fix Coverage

srijanreddy98 commented on core/templates/domain/collection/collection-node.model.ts:
```diff
@@ -41,6 +38,19 @@ export class CollectionNode {
+    return this.create({
+      exploration_id: explorationId,
+      exploration_summary: null
+    });
```
static functions don't have access to the `this` keyword. This refers to the instance of the model but static functions are bound to the class and not the instance. So this will become:
```suggestion
    return CollectionNode.create({
      exploration_id: explorationId,
      exploration_summary: null
    });
```

darksun27 replied:
Done here and elsewhere!

srijanreddy98 commented on core/templates/domain/collection/collection-update.service.spec.ts:
```diff
@@ -81,8 +69,8 @@ describe('Collection update service', function() {
-    _sampleCollection = collectionObjectFactory.create(
-      sampleCollectionBackendObject);
+    _sampleCollection = Collection.create(
+      sampleCollectionBackendObject as unknown as CollectionBackendDict);
```
why `as unknown as CollectionBackendDict`? Can this just be `as CollectionBackendDict`? Alternatively on line 55, change `var sampleCollectionBackendObject` to `const sampleCollectionBackendObject: CollectionBackendDict`

srijanreddy98 replied:
After taking a look at `sampleCollectionBackendObject`, it was not of the type "CollectionBackendDict". And the type mismatch (empty object for exploration summary) doesn't match the type LearnerExplorationSummaryBackendDict. As this was going really deep, I am ok with this for now. Can you create an issue and assign it to me.

darksun27 replied:
Done assigned you!

srijanreddy98 replied:
Sorry, I forgot to mention, add a TODO before this line as well. `TODO(#10875): Fix type mismatch`. here and the other 4 places in this or you have used.

darksun27 replied:
Done!

srijanreddy98 commented on core/templates/domain/email-dashboard/email-dashboard-query-results.model.ts:
```diff
@@ -17,14 +17,10 @@
  * results.
```
File overview not changed for this file

darksun27 replied:
Done!

srijanreddy98 commented on core/templates/domain/email-dashboard/email-dashboard-query.model.ts:
```diff
@@ -13,12 +13,9 @@
 
 /**
- * @fileoverview Frontend domain object factory for email dashboard query.
+ * @fileoverview Frontend domain model for email dashboard query.
```
*Frontend Model (don't use domain)

darksun27 replied:
Done!

srijanreddy98 commented on core/templates/domain/improvements/ineffective-feedback-loop-task.model.spec.ts:
```diff
@@ -16,23 +16,12 @@
  * @fileoverview Unit tests for the IneffectiveFeedbackLoopTask domain object.
```
Domain Object - Model

srijanreddy98 commented on core/templates/domain/improvements/ineffective-feedback-loop-task.model.ts:
```diff
@@ -17,13 +17,8 @@
  *    task.
```
Change fileoverview here.

darksun27 replied:
Done!

srijanreddy98 commented on core/templates/domain/improvements/ineffective-feedback-loop-task.model.ts:
```diff
@@ -90,7 +62,7 @@ export class IneffectiveFeedbackLoopTaskObjectFactory {
   }
 
-  createNew(
+  static createNew(
```
don't use "this" inside static functions

darksun27 replied:
Done!

srijanreddy98 commented on core/templates/domain/improvements/needs-guiding-response-task.model.ts:
```diff
@@ -94,21 +64,39 @@ export class NeedsGuidingResponsesTaskObjectFactory {
   }
 
-  createFromAnswerStats(
+  static createFromAnswerStats(
```
Don't use this inside static functions.

Can you check for this pattern in all the other files you have changed (and also in your previous PR)?

darksun27 replied:
Done!

srijanreddy98 reviewed (changes requested):
@darksun27 - requested some changes. Please take a look.

darksun27 committed:
Fix Merge Conflicts

darksun27 committed:
Fix Review Issues

darksun27 committed:
Fix Frontend Test Issues

darksun27 commented:
@srijanreddy98 PTAL!

srijanreddy98 reviewed (changes requested):
Nicely done @darksun27. One minor nit otherwise LGTM

darksun27 committed:
Add TODO(#10875): Fix type mismatch

darksun27 committed:
Add TODO(#10875): Fix type mismatch

darksun27 commented:
@

> Nicely done @darksun27. One minor nit otherwise LGTM

@srijanreddy98 Done! PTAL

vojtechjelinek reviewed (approved):
LGTM!

DubeySandeep reviewed (approved):
LGTM for the 3 codeowner files*

darksun27 committed:
Fix Merge Conflicts

BenHenning commented on core/templates/domain/platform_feature/platform-parameter-filter.model.ts:
```diff
@@ -13,14 +13,11 @@
 
 /**
- * @fileoverview Factory for creating PlatformParameterFilter domain objects.
+ * @fileoverview Frontend Model for Platform Parameter Filter.
```
I suggest: ``Frontend model for platform parameter filters.``

The previous wording didn't quite read correctly. Ditto elsewhere in this PR (where applicable).

BenHenning commented on core/templates/domain/platform_feature/platform-parameter-rule.model.ts:
```diff
@@ -13,18 +13,14 @@
 
 /**
- * @fileoverview Factory for creating PlatformParameterRule domain objects.
+ * @fileoverview Frontend Model for Platform Parameter Rule.
```
Ditto here (documentation wording) & elsewhere in the PR.

BenHenning reviewed (approved):
Thanks @darksun27! This is a really nice improvement.

Approved for code owners. I left a couple of general documentation wording suggestions.

darksun27 committed:
Merge remote-tracking branch 'upstream/develop' into refactor-object-factories-red

vinitamurthi reviewed (approved):
LGTM as a codeowner

kevintab95 commented on core/templates/domain/collection/collection-node.model.spec.ts:
```diff
@@ -13,19 +13,13 @@
 
 /**
- * @fileoverview Tests for CollectionNodeObjectFactory.
+ * @fileoverview Tests for collection-node.model.
```
```suggestion
 * @fileoverview Tests for CollectionNodeModel.
```

So that it matches the pattern used in other files.

kevintab95 commented on core/templates/domain/collection/CollectionObjectFactorySpec.ts:
```diff
@@ -13,29 +13,18 @@
 
 /**
- * @fileoverview Tests for CollectionObjectFactory.
+ * @fileoverview Tests for collection-model.
```
```suggestion
 * @fileoverview Tests for CollectionModel.
```

So that it matches other files.

kevintab95 reviewed (approved):
Thanks a lot @darksun27! This is a nice simplification. LGTM with 2 nits.

DubeySandeep commented:
@darksun27, Any update on this PR? 

@Showtim3 @brianrodri CAn you please review this PR?

darksun27 committed:
Fix Conflicts:
Fixed merge conflicts in the following files

- core/templates/components/oppia-angular-root.component.ts
- core/templates/domain/admin/admin-backend-api.service.spec.ts
- core/templates/domain/admin/admin-backend-api.service.ts
- core/templates/domain/improvements/high-bounce-rate-task.model.spec.ts
- core/templates/domain/learner_dashboard/learner-dashboard-backend-api.service.ts
- core/templates/domain/topics_and_skills_dashboard/topics-and-skills-dashboard-backend-api.service.ts
- core/templates/pages/admin-page/services/admin-data.service.spec.ts
- core/templates/pages/learner-dashboard-page/learner-dashboard-page.component.spec.ts
- core/templates/pages/topics-and-skills-dashboard-page/topics-and-skills-dashboard-page.component.spec.ts
- core/templates/services/UpgradedServices.ts
- core/templates/services/angular-services.index.ts
- core/templates/services/exploration-improvements-backend-api.service.ts
- core/templates/services/exploration-improvements.service.spec.ts

darksun27 committed:
Fix Merge conflicts in angular-services

darksun27 committed:
Fix compile issues

Showtim3 reviewed (approved):
LGTM
